cayenne-server/src/main/java/org/apache/cayenne/util/ObjectContextQueryAction.java

```diff
@@ -369,7 +370,21 @@ protected boolean interceptLocalCache() {
             // response may already be initialized by the factory above ... it is null if
             // there was a preexisting cache entry
             if (response == null) {
-                response = new ListResponse(cachedResults);
+                response = new ListResponse(cachedResults != null ? Collections.unmodifiableList(cachedResults) : null); // make a defensive copy avoid leaking cache internal data
```

Comment:
I think we should store an immutable list in the cache, not wrpa it when we retrieve it.